File: backend/cr8_engine/app/schemas/moodboard.py
from typing import List, Optional
from pydantic import BaseModel
from datetime import datetime


class MoodboardBase(BaseModel):
    user_id: int
    name: str
    description: str
    storyline: Optional[str] = None
    keywords: List[str] = []  # Default empty list (not Optional)
    industry: Optional[str] = None
    targetAudience: Optional[str] = None
    theme: Optional[str] = None
    tone: Optional[str] = None
    videoReferences: List[str] = []  # Default empty list
    colorPalette: List[str] = []  # Default empty list
    usage_intent: Optional[str] = None

    class Config:
        orm_mode = True


class MoodboardCreate(MoodboardBase):
    pass


class MoodboardUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    storyline: Optional[str] = None
    keywords: Optional[List[str]] = None
    industry: Optional[str] = None
    targetAudience: Optional[str] = None
    theme: Optional[str] = None
    tone: Optional[str] = None
    videoReferences: Optional[List[str]] = None
    colorPalette: Optional[List[str]] = None
    usage_intent: Optional[str] = None

    class Config:
        orm_mode = True


class MoodboardRead(MoodboardBase):
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True

File: frontend/lib/types/moodboard.ts
import { z } from "zod";

export type Theme =
  | "minimalist"
  | "futuristic"
  | "rustic"
  | "industrial"
  | "vibrant"
  | "custom";
export type Tone =
  | "joyful"
  | "melancholic"
  | "energetic"
  | "serene"
  | "bold"
  | "dreamy";
export type Industry =
  | "fashion"
  | "music"
  | "real-estate"
  | "tech"
  | "entertainment"
  | "custom";
export type UsageIntent =
  | "product-launch"
  | "music-video"
  | "social-media"
  | "commercial";

export interface ImageWithAnnotation {
  file: File;
  annotation?: string;
}

export interface CategoryImages {
  compositions: ImageWithAnnotation[];
  actions: ImageWithAnnotation[];
  lighting: ImageWithAnnotation[];
  location: ImageWithAnnotation[];
}

export interface MoodboardData {
  id: number;
  name: string;
  description: string;
  user_id: number;
  categoryImages: CategoryImages;
  colorPalette: string[];
  videoReferences: string[];
  theme: Theme;
  storyline: string;
  keywords: string[];
  tone: Tone;
  industry: Industry;
  targetAudience: string;
  usageIntent: UsageIntent;
}

// Zod Types

// Define the file schema if not already defined
const fileSchema = z.object({
  file: z.instanceof(File),
  annotation: z.string().optional(),
});

export const moodboardSchema = z.object({
  categoryImages: z.object({
    compositions: z
      .array(fileSchema)
      .min(1, "Add at least one composition reference image"),
    actions: z
      .array(fileSchema)
      .min(1, "Add at least one action reference image"),
    lighting: z
      .array(fileSchema)
      .min(1, "Add at least one lighting reference image"),
    location: z
      .array(fileSchema)
      .min(1, "Add at least one location reference image"),
  }),
  colorPalette: z
    .array(z.string())
    .min(1, "Add at least one color to the palette"),
  videoReferences: z.array(z.string().url("Please enter a valid URL")),
  theme: z.enum(
    [
      "minimalist",
      "futuristic",
      "rustic",
      "industrial",
      "vibrant",
      "custom",
    ] as const,
    {
      required_error: "Please select a theme",
    }
  ),
  storyline: z.string().min(1, "Please enter a storyline"),
  keywords: z.array(z.string()).min(1, "Add at least one keyword"),
  tone: z.enum(
    ["joyful", "melancholic", "energetic", "serene", "bold", "dreamy"] as const,
    {
      required_error: "Please select a tone",
    }
  ),
  industry: z.enum(
    [
      "fashion",
      "music",
      "real-estate",
      "tech",
      "entertainment",
      "custom",
    ] as const,
    {
      required_error: "Please select an industry",
    }
  ),
  targetAudience: z.string().min(1, "Please specify the target audience"),
  usageIntent: z.enum(
    ["product-launch", "music-video", "social-media", "commercial"] as const,
    {
      required_error: "Please select a usage intent",
    }
  ),
});

export type MoodboardFormData = z.infer<typeof moodboardSchema>;

export interface MoodboardList extends MoodboardFormData {
  id: number;
  name: string;
  description: string;
  user_id: number;
}
